core/modules/qdisp/ResponseRequester.h

```diff
@@ -101,8 +109,10 @@ class ResponseRequester {
         boost::shared_ptr<CancelFunc> f;
         {
             boost::lock_guard<boost::mutex> lock(_cancelMutex);
-            f.swap(_cancelFunc);
-            _cancelled = true;
+            if(!_cancelled) {
+                f = _cancelFunc;
```

Comment:
I tried that, but I backed off, because for some reason it really upsets Qserv badly to the point it segfaults almost every time. Hmm.